retrieval/scrape.go

```diff
@@ -446,6 +461,28 @@ func (sl *scrapeLoop) stop() {
 	<-sl.done
 }
 
+func (sl *scrapeLoop) processScrapeResult(samples model.Samples, scrapeErr error, start time.Time) error {
+	// Collect samples post-relabelling and label handling in a buffer.
+	buf := &bufferAppender{buffer: make(model.Samples, 0, len(samples))}
```

Comment:
Heap profiling tells you about the current state of the heap. If GC runs more frequently than the scrape cycle, you will not see all the accumulated buffer allocations of a full scrape cycle but only those that happened since the last GC run.
It would be interesting to see how much CPU is used in GC before and after.

> It's not easy to avoid, we'd need two significantly different code paths.

But the one is just pass-through with the only slight complication of counting the number of post-metric relabelling samples.

I might not see all implications right now. As said, spread thinly on time right now to do a thorough analysis. I don't want to block this if other find the approach OK. But I cannot convince myself to find it OK right now. I'd recommend to just play safe and implement the simple code path for the unlimited case.

Perhaps @juliusv or @fabxc can chime in and confirm or reject my intuition? Happy to accept any outcome as I don't have time for a lengthy discussion either.